core_data_modules/traced_data/io.py

```diff
@@ -439,6 +445,305 @@ def import_coda_to_traced_data_iterable_as_matrix(cls, user, data, key_of_raw, c
             td.append_data(td_matrix_data, Metadata(user, Metadata.get_call_location(), time.time()))
 
 
+class TracedDataCoda2IO(object):
+    @classmethod
+    def add_message_ids(cls, user, data, raw_key, message_id_key):
+        """
+        Appends a message id to each object in the provided iterable of TracedData.
+
+        Message ids are set by computing the SHA of the value at each `raw_key`, so are guaranteed to be stable.
+
+        If the `raw_key` is not found in a TracedData object in the iterable, no message id is assigned.
+
+        :param user: Identifier of the user running this program, for TracedData Metadata.
+        :type user: str
+        :param data: TracedData objects to set message_ids for.
+        :type data: iterable of TracedData
+        :param raw_key: Key in TracedData objects to read the text to generate message ids for from.
+        :type raw_key: str
+        :param message_id_key: Key in TracedData objects to write the message id to.
+        :type message_id_key: str
+        """
+        for td in data:
+            if raw_key in td:
+                td.append_data(
+                    {message_id_key: SHAUtils.sha_string(td[raw_key])},
+                    Metadata(user, Metadata.get_call_location(), time.time())
+                )
+
+    @staticmethod
+    def _assert_uniquely_coded(data, message_id_key, coded_keys):
+        """
+        Checks that all objects with the same message id have been assigned the same codes for each coded_key in
+        coded_keys.
+
+        Fails with an AssertionError if there are TracedData objects with the same message id but different codes
+        assigned, otherwise has no side-effects.
```

Comment:
Does Python have anything like an @ pure annotation?

Reply:
not built in